File: natlang/trace.py
"""Versioned, append-only execution observations and an offline reader."""
from __future__ import annotations

import json
import copy
import threading
from pathlib import Path
from typing import Any


VERSION = "reduction-trace/1"


def _view(value):
    if value is None or isinstance(value, (str, bool, int)):
        return value
    if isinstance(value, float):
        return value if value == value and abs(value) != float("inf") else {"$opaque": "nonfinite-number"}
    if isinstance(value, (list, tuple)):
        return [_view(v) for v in value]
    if isinstance(value, dict):
        return {str(k): _view(v) for k, v in value.items()}
    return {"$opaque": type(value).__name__, "reconstructable": False}


def _type_name(value):
    if value is None:
        return "Null"
    if isinstance(value, bool):
        return "Bool"
    if isinstance(value, (int, float)):
        return "Num"
    if isinstance(value, str):
        return "Text"
    if isinstance(value, list):
        return "List"
    if isinstance(value, dict):
        for body in value.values():
            if isinstance(body, dict) and "type" in body and any(k.startswith("$") for k in value):
                return body["type"]
        return "Record"
    return "Opaque"


def changes(before, after, path=()):
    """Typed, presence-aware changes between portable state snapshots."""
    if isinstance(before, dict) and isinstance(after, dict):
        result = []
        for key in sorted(before.keys() | after.keys(), key=str):
            if key not in before or key not in after:
                result.append({"path": [*path, key], "before_present": key in before,
                               "after_present": key in after,
                               "before": before.get(key), "after": after.get(key),
                               "type": _type_name(after.get(key)) if key in after else None})
            else:
                result.extend(changes(before[key], after[key], (*path, key)))
        return result
    if before == after:
        return []
    return [{"path": list(path), "before_present": True, "after_present": True,
             "before": before, "after": after, "type": _type_name(after)}]


class TraceRecorder:
    def __init__(self, manifest: dict, path: Path | None = None):
        self.events: list[dict] = []
        self._lock = threading.Lock()
        self.path = Path(path) if path is not None else None
        self._stream = self.path.open("w", encoding="utf-8") if self.path else None
        self.emit("manifest", **manifest)

    def emit(self, kind: str, **data) -> dict:
        with self._lock:
            event = _view({"version": VERSION, "seq": len(self.events), "kind": kind, **data})
            encoded = json.dumps(event, ensure_ascii=False, allow_nan=False)
            clean = json.loads(encoded)
            self.events.append(clean)
            if self._stream:
                self._stream.write(encoded + "\n")
                self._stream.flush()
            return clean

    def close(self):
        if self._stream:
            self._stream.close()
            self._stream = None

    def reopen(self):
        if self.path is not None and self._stream is None:
            self._stream = self.path.open("a", encoding="utf-8")


class TraceReader:
    def __init__(self, events: list[dict]):
        if not events or events[0].get("kind") != "manifest":
            raise ValueError("trace has no manifest")
        if any(event.get("version") != VERSION or event.get("seq") != i
               for i, event in enumerate(events)):
            raise ValueError("unsupported or discontinuous trace")
        self.events = events

    @classmethod
    def open(cls, path: Path):
        with Path(path).open(encoding="utf-8") as stream:
            return cls([json.loads(line) for line in stream if line.strip()])

    @property
    def manifest(self) -> dict:
        return self.events[0]

    def of_kind(self, kind: str) -> list[dict]:
        return [event for event in self.events if event["kind"] == kind]

    def final_state(self) -> Any:
        snapshots = self.of_kind("state")
        if not snapshots:
            raise ValueError("trace has no captured state")
        return snapshots[-1]["value"]

    def reconstruct(self) -> Any:
        states = self.of_kind("state")
        if not states or states[0].get("phase") != "initial":
            raise ValueError("trace has no initial state")
        current = copy.deepcopy(states[0]["value"])
        for event in self.of_kind("reduction"):
            for change in event["changes"]:
                path = change["path"]
                if not path:
                    current = copy.deepcopy(change["after"])
                    continue
                target = current
                for key in path[:-1]:
                    target = target[key]
                if change["after_present"]:
                    target[path[-1]] = copy.deepcopy(change["after"])
                else:
                    del target[path[-1]]
        if current != states[-1]["value"]:
            raise ValueError("reduction changes do not reconstruct the final state")
        return current

    def coverage(self) -> dict:
        snapshots = self.of_kind("state")
        incomplete = any("$stream" in json.dumps(event.get("value")) or
                         '"$opaque"' in json.dumps(event.get("value")) for event in snapshots)
        return {"state_reconstructable": bool(self.of_kind("state")),
                "live_source_reconstructable": not incomplete,
                "native_effects_replayable": False,
                "effect_count": len(self.of_kind("effect")),
                "mode": "recorded-observations-only"}

    def replay_observations(self) -> dict:
        """Read captured decisions and state; never call model, eval or host effects."""
        states = self.of_kind("state")
        if not states or states[0].get("phase") != "initial" or states[-1].get("phase") != "final":
            raise ValueError("trace lacks complete initial/final observations")
        return {"initial": states[0]["value"], "final": states[-1]["value"],
                "outcome": states[-1].get("outcome"),
                "actions": self.of_kind("action"), "effects": self.of_kind("effect"),
                "coverage": self.coverage()}

File: ts-host/src/native/trace.ts
import { digest } from './hash.js';

export const TRACE_VERSION = 'reduction-trace/1';
export type TraceEvent = { version: typeof TRACE_VERSION; seq: number; kind: string; [key: string]: unknown };

export function deriveSeed(root: number, path: string, attempt: number, purpose: string, ordinal = 0): number {
  const payload = { attempt, ordinal, path, purpose, root, version: 'sha256-json-v1' };
  const bytes = digest(JSON.stringify(payload));
  let first = 0n;
  for (const byte of bytes.slice(0, 8)) first = (first << 8n) | BigInt(byte);
  return Number(first % (2n ** 31n));
}

function valueType(value: unknown): string {
  if (value === null) return 'null';
  if (typeof value === 'boolean') return 'boolean';
  if (typeof value === 'number') return 'number';
  if (typeof value === 'string') return 'string';
  if (Array.isArray(value)) return 'List';
  if (value && typeof value === 'object') for (const [key, body] of Object.entries(value)) {
    if (key.startsWith('$') && body && typeof body === 'object' && 'type' in body)
      return String((body as Record<string, unknown>).type);
  }
  return 'Record';
}

function canonical(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === 'object')
    return Object.fromEntries(Object.entries(value).sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0)
      .map(([key, item]) => [key, canonical(item)]));
  return value;
}
const sameValue = (a: unknown, b: unknown) => JSON.stringify(canonical(a)) === JSON.stringify(canonical(b));

export function changes(before: unknown, after: unknown, path: (string | number)[] = []): Record<string, unknown>[] {
  if (before && after && typeof before === 'object' && typeof after === 'object' &&
      !Array.isArray(before) && !Array.isArray(after)) {
    const a = before as Record<string, unknown>, b = after as Record<string, unknown>;
    return [...new Set([...Object.keys(a), ...Object.keys(b)])].sort().flatMap(key => {
      if (!(key in a) || !(key in b)) return [{ path: [...path, key], before_present: key in a,
        after_present: key in b, before: a[key] ?? null, after: b[key] ?? null,
        type: key in b ? valueType(b[key]) : null }];
      return changes(a[key], b[key], [...path, key]);
    });
  }
  if (sameValue(before, after)) return [];
  return [{ path, before_present: true, after_present: true, before, after, type: valueType(after) }];
}

export class NativeTraceRecorder {
  readonly events: TraceEvent[] = [];
  private readonly started = performance.now();
  constructor(manifest: Record<string, unknown>) { this.emit('manifest', manifest); }
  /** Open a captured trace for read-only reconstruction and admission checks. */
  static fromEvents(events: Record<string, unknown>[]): NativeTraceRecorder {
    const trace = new NativeTraceRecorder({});
    trace.events.splice(0, trace.events.length, ...structuredClone(events) as TraceEvent[]);
    trace.reconstruct();
    return trace;
  }
  emit(kind: string, data: Record<string, unknown> = {}): TraceEvent {
    const event = JSON.parse(JSON.stringify({ version: TRACE_VERSION, seq: this.events.length, kind,
      observed_at: new Date().toISOString(), elapsed_ms: Math.round(performance.now() - this.started),
      ...data })) as TraceEvent;
    this.events.push(event);
    return event;
  }
  finalState(): unknown {
    const states = this.events.filter(event => event.kind === 'state');
    if (!states.length) throw new Error('trace has no state');
    return states.at(-1)!.value;
  }
  coverage(): Record<string, unknown> {
    const states = this.events.filter(event => event.kind === 'state');
    const incomplete = states.some(event => /"\$stream"|"\$opaque"/.test(JSON.stringify(event.value)));
    return { state_reconstructable: states.length > 0, live_source_reconstructable: !incomplete,
      native_effects_replayable: false, effect_count: this.events.filter(event => event.kind === 'effect').length,
      mode: 'recorded-observations-only' };
  }
  replayObservations(): Record<string, unknown> {
    const states = this.events.filter(event => event.kind === 'state');
    if (states[0]?.phase !== 'initial' || states.at(-1)?.phase !== 'final')
      throw new Error('trace lacks complete initial/final observations');
    return { initial: states[0]!.value, final: states.at(-1)!.value, outcome: states.at(-1)!.outcome,
      actions: this.events.filter(event => event.kind === 'action'),
      effects: this.events.filter(event => event.kind === 'effect'), coverage: this.coverage() };
  }
  reconstruct(): unknown {
    if (this.events[0]?.kind !== 'manifest' || this.events.some((event, index) =>
      event.version !== TRACE_VERSION || event.seq !== index)) throw new Error('unsupported or discontinuous trace');
    const first = this.events.find(event => event.kind === 'state' && event.phase === 'initial');
    if (!first) throw new Error('trace has no initial state');
    let current = structuredClone(first.value);
    for (const event of this.events.filter(event => event.kind === 'reduction')) {
      for (const change of event.changes as Record<string, unknown>[]) {
        const path = change.path as (string | number)[];
        if (!path.length) { current = structuredClone(change.after); continue; }
        let target = current as Record<string, unknown>;
        for (const part of path.slice(0, -1)) target = target[String(part)] as Record<string, unknown>;
        if (change.after_present) target[String(path.at(-1))] = structuredClone(change.after);
        else delete target[String(path.at(-1))];
      }
    }
    if (!sameValue(current, this.finalState())) throw new Error('trace reconstruction mismatch');
    return current;
  }
}
